src/Uno.UI.Runtime.Skia/GLCanvasElement.cs

```diff
@@ -0,0 +1,190 @@
+
+	private readonly GLVisual _visual;
+
+	private GL? _gl;
+	private uint _framebuffer;
+	private uint _textureColorBuffer;
+	private unsafe readonly void* _pixels;
+	private uint _renderBuffer;
+
+	unsafe protected GLCanvasElement(Size resolution)
+	{
+		_width = (uint)resolution.Width;
+		_height = (uint)resolution.Height;
+		_pixels = (void*)Marshal.AllocHGlobal((int)(_width * _height * BytesPerPixel));
+
+		_visual = new GLVisual(this, Visual.Compositor);
+		Visual.Children.InsertAtTop(_visual);
+	}
+
+	unsafe ~GLCanvasElement()
+	{
+		Marshal.FreeHGlobal((IntPtr)_pixels);
+
+		if (_gl is { })
+		{
+			_gl.DeleteFramebuffer(_framebuffer);
+			_gl.DeleteTexture(_textureColorBuffer);
+			_gl.DeleteRenderbuffer(_renderBuffer);
+		}
+	}
+
+	protected abstract void Init(GL gl);
+	protected abstract void OnDestroy(GL gl);
+	protected abstract void RenderOverride(GL gl);
+
+	public void Invalidate() => _visual.Compositor.InvalidateRender(_visual);
+
+	private unsafe protected override void OnLoaded()
+	{
+		base.OnLoaded();
+
+		ApiExtensibility.CreateInstance<GLGetProcAddress>(this, out var getProcAddress);
+		getProcAddress = getProcAddress ?? throw new InvalidOperationException($"Couldn't get GetProcAddress for {nameof(GLCanvasElement)}. Make sure you are running on a platform with {nameof(GLCanvasElement)} support.");
+		_gl = GL.GetApi(getProcAddress.Invoke);
+
+		if (_firstLoad)
+		{
+			_firstLoad = false;
+
+			using var _ = new GLStateDisposable(_gl);
+
+			_framebuffer = _gl.GenBuffer();
+			_gl.BindFramebuffer(GLEnum.Framebuffer, _framebuffer);
+			{
+				_textureColorBuffer = _gl.GenTexture();
+				_gl.BindTexture(GLEnum.Texture2D, _textureColorBuffer);
+				{
+					_gl.TexImage2D(GLEnum.Texture2D, 0, InternalFormat.Rgb, _width, _height, 0, GLEnum.Rgb, GLEnum.UnsignedByte, (void*)0);
+					_gl.TexParameterI(GLEnum.Texture2D, GLEnum.TextureMinFilter, (uint)GLEnum.Linear);
+					_gl.TexParameterI(GLEnum.Texture2D, GLEnum.TextureMagFilter, (uint)GLEnum.Linear);
+					_gl.FramebufferTexture2D(GLEnum.Framebuffer, FramebufferAttachment.ColorAttachment0, GLEnum.Texture2D, _textureColorBuffer, 0);
+				}
+				_gl.BindTexture(GLEnum.Texture2D, 0);
+
+				_renderBuffer = _gl.GenRenderbuffer();
+				_gl.BindRenderbuffer(GLEnum.Renderbuffer, _renderBuffer);
+				{
+					_gl.RenderbufferStorage(GLEnum.Renderbuffer, InternalFormat.Depth24Stencil8, _width, _height);
+					_gl.FramebufferRenderbuffer(GLEnum.Framebuffer, GLEnum.DepthStencilAttachment, GLEnum.Renderbuffer, _renderBuffer);
+
+					Init(_gl);
+				}
+				_gl.BindRenderbuffer(GLEnum.Renderbuffer, 0);
+
+				if (_gl.CheckFramebufferStatus(GLEnum.Framebuffer) != GLEnum.FramebufferComplete)
+				{
+					throw new InvalidOperationException("Offscreen framebuffer is not complete");
+				}
+			}
+			_gl.BindFramebuffer(GLEnum.Framebuffer, 0);
+		}
+
+		Render();
+	}
+
+	private unsafe void Render()
+	{
+		if (!IsLoaded)
+		{
+			return;
+		}
+
+		using var _ = new GLStateDisposable(_gl!);
```

Comment:
What if `_gl` is null?

Reply:
_gl cannot be be null if the element is `IsLoaded`, since we attempt to create a GL object there and throw if we couldn't